pkg/fileutils/filelock.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fileutils
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"syscall"
+)
+
+// FileLock define a file lock implemented by syscall.Flock
+type FileLock struct {
+	fileName string
+	fd       *os.File
+	mutex    sync.Mutex
+}
+
+// NewFileLock create a FileLock instance
+func NewFileLock(name string) *FileLock {
+	return &FileLock{
+		fileName: name,
+	}
+}
+
+// Lock locks file.
+// If the file is already locked, the calling goroutine blocks until the file is unlocked.
+// If lock has been invoked without unlock, lock again will return an error.
+func (l *FileLock) Lock() error {
+	fd, err := os.Open(l.fileName)
+	if err != nil {
+		return err
+	}
+
+	l.mutex.Lock()
+	if l.fd != nil {
+		l.mutex.Unlock()
+		return fmt.Errorf("invoke Unlock() before lock %s again", l.fileName)
+	}
+	l.fd = fd
+	l.mutex.Unlock()
+
+	err = syscall.Flock(int(l.fd.Fd()), syscall.LOCK_EX)
+	if err != nil {
+		return fmt.Errorf("file %s lock failed error: %s", l.fileName, err)
```

Comment:
maybe it's better to use errors.Wrapf(err,"file %s lock failed",l.fileName) 😄

Reply:
I have wrapped the errors in lock&unlock.